examples/asset/multi_asset_sync.rs

```diff
@@ -0,0 +1,296 @@
+
+/// Tracks how many guards are remaining.
+#[derive(Debug, Resource)]
+pub struct AssetBarrierInner {
+    count: AtomicU32,
+    /// This can be omitted if async is not needed.
+    notify: Event,
+}
+
+/// State of loading asynchronously.
+#[derive(Debug, Resource)]
+pub struct AsyncLoadingState(Arc<AtomicBool>);
+
+/// Entities that are to be removed once loading finished
+#[derive(Debug, Component)]
+pub struct Loading;
+
+/// Marker for the "Loading..." Text component.
+#[derive(Debug, Component)]
+pub struct LoadingText;
+
+impl AssetBarrier {
+    /// Create an [`AssetBarrier`] with a [`AssetBarrierGuard`].
+    pub fn new() -> (AssetBarrier, AssetBarrierGuard) {
+        let inner = Arc::new(AssetBarrierInner {
+            count: AtomicU32::new(1),
+            notify: Event::new(),
+        });
+        (AssetBarrier(inner.clone()), AssetBarrierGuard(inner))
+    }
+
+    /// Returns true if all [`AssetBarrierGuard`] is dropped.
+    pub fn is_ready(&self) -> bool {
+        self.count.load(Ordering::Acquire) == 0
+    }
+
+    /// Wait for all [`AssetBarrierGuard`]s to be dropped asynchronously.
+    pub fn wait_async(&self) -> impl Future<Output = ()> + 'static {
+        let shared = self.0.clone();
+        async move {
+            loop {
+                // Acquire an event listener.
+                let listener = shared.notify.listen();
+                // If all barrier guards are dropped, return
+                if shared.count.load(Ordering::Acquire) == 0 {
+                    return;
+                }
+                // Wait for the last barrier guard to notify us
+                listener.await;
+            }
+        }
+    }
+}
+
+// Increment count on clone.
+impl Clone for AssetBarrierGuard {
+    fn clone(&self) -> Self {
+        self.count.fetch_add(1, Ordering::AcqRel);
+        AssetBarrierGuard(self.0.clone())
+    }
+}
+
+// Decrement count on clone.
+impl Drop for AssetBarrierGuard {
+    fn drop(&mut self) {
+        let prev = self.count.fetch_sub(1, Ordering::AcqRel);
+        if prev == 1 {
+            // Notify all listeners if count reaches 0.
+            self.notify.notify(usize::MAX);
+        }
+    }
+}
+
+fn main() {
+    App::new()
+        .add_plugins(DefaultPlugins)
+        .init_state::<LoadingState>()
+        .insert_resource(AmbientLight {
+            color: Color::WHITE,
+            brightness: 2000.,
+        })
+        .add_systems(Startup, setup)
+        // This showcases how to wait for assets synchronously.
+        .add_systems(Update, wait_on_load)
+        // This showcases how to wait for assets asynchronously.
+        .add_systems(
+            Update,
+            get_async_loading_state.run_if(in_state(LoadingState::Loading)),
+        )
+        // This showcases how to react to asynchronous world mutation synchronously.
+        .add_systems(
+            OnExit(LoadingState::Loading),
+            despawn_loading_state_entities,
+        )
+        .run();
+}
```

Comment:
Because this example is fairly lengthy, I think it would be good to move this main function to the top, since it gives a pretty good overview of what you're trying to achieve in ~20 LOC.

Reply:
done!